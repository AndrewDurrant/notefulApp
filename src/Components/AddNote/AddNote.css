.addNote {
  margin: 2rem;
  display: flex;
  justify-content: center;
  flex-direction: column;
}