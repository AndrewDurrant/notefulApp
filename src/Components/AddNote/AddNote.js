import React, { Component } from 'react';
import cuid from 'cuid';
import ValidationError from '../../ValidationError';
import NotefulContext from '../../NotefulContext';
import './AddNote.css'
import { Link } from 'react-router-dom';


export class AddNote extends Component {
  static contextType = NotefulContext;

  constructor(props) {
    super(props);
    this.state = {
      noteName: {
        value: '',
        touched: false
      },
      noteContent: {
        value: '',
        touched: false
      },
      folderId: {
        value: '',
        touched: false
      }
    }
  }

  handleChange(folder) {
    this.setState({
      folderId: {
        value: folder,
        touched: true
      }
    });
  }

  updateNoteName(name) {
    this.setState({
      noteName: {
        value: name,
        touched: true
      }
    });
  }

  updateNoteContent(content) {
    this.setState({
      noteContent: {
        value: content,
        touched: true
      }
    })
  }

  handleSubmit(event) {
    event.preventDefault();
    const { noteName, noteContent, folderId } = this.state;
    const note = {
      id: cuid(),
      folderId: folderId.value,
      name: noteName.value,
      content: noteContent.value
    }
    this.context.addNote(note)
      .then(() => {
        // add note
        this.context.updateNotes(note)
        this.props.history.push('/')
      })
  }

  validateNoteName() {
    const noteName = this.state.noteName.value.trim();
    if(noteName.length === 0) {
      return 'Note name is required';
    } else if(noteName.length < 3) {
      return 'Note name must be at least 3 characters long'
    }
  }

  validateNoteContent() {
    const noteContent = this.state.noteContent.value.trim();
    if(noteContent.length === 0) {
      return 'Note content is required';
    } else if(noteContent.length < 3) {
      return 'Note content must be at least 7 characters long'
    }
  }

  render() {
    const noteNameError = this.validateNoteName();
    const noteContentError = this.validateNoteContent();

    return(
      <form className="addNote" onSubmit={e => this.handleSubmit(e)}>
        <h2>Add Note</h2>
        <span>* required field</span>
        <section>
          <label htmlFor="noteName">Note Name *</label>
          <input 
            type="text"
            name="noteName"
            id="noteName"
            onChange={e => this.updateNoteName(e.target.value)}
          />
          {this.state.noteName.touched && (
            <ValidationError message={noteNameError} />
          )}
        </section>

        <section>
          <label htmlFor="noteContent">Note Content *</label>
          <textarea 
            type="text"
            name="noteContent"
            id="noteContent"
            placeholder="Note it up!"
            cols="66"
            rows="3"
            onChange={e => this.updateNoteContent(e.target.value)}
          />
          {this.state.noteContent.touched && (
            <ValidationError message={noteContentError} />
          )}
        </section>

        <section>
          <label htmlFor="folderId">Folder Association</label>
          <select 
            name="folderId" 
            id="folderId" 
            value={this.state.value} 
            onChange={e => this.handleChange(e.target.value)}
          >
            {this.context.folders.map(folder =>
              <option value={folder.id} key={folder.id}>{folder.name}</option>
            )}
          </select>
        </section>

        <section className="addNote__button__group">
          <Link to='/'>
            <button type="reset">
              Cancel
            </button>
          </Link>
          <button
            type="submit"
            disabled= {
              this.validateNoteName() ||
              this.validateNoteContent()
            }
          >
            Save
          </button>
        </section>
      </form>
    )
  }

}

export default AddNote