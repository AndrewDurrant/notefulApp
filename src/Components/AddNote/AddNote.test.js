import React from 'react';
import ReactDOM from 'react-dom';
import AddNote from './AddNote.js';
import { MemoryRouter } from 'react-router-dom';
import renderer from 'react-test-renderer';

it('renders without crashing', () => {
  const div = document.createElement('div');

  ReactDOM.render(
    <MemoryRouter>
      <AddNote />
    </MemoryRouter>
    , div
  );

  ReactDOM.unmountComponentAtNode(div);
})

it('renders the UI as expected', () => {
  const tree = renderer
    .create(<MemoryRouter><AddNote /></MemoryRouter>)
    .toJSON();
  expect(tree).toMatchSnapshot();
})