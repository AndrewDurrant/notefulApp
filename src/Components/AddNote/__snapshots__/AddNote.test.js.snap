// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders the UI as expected 1`] = `
<form
  className="addNote"
  onSubmit={[Function]}
>
  <h2>
    Add Note
  </h2>
  <div>
    * required field
  </div>
  <div>
    <label
      htmlFor="noteName"
    >
      Note Name *
    </label>
    <input
      id="noteName"
      name="noteName"
      onChange={[Function]}
      type="text"
    />
  </div>
  <div>
    <label
      htmlFor="noteContent"
    >
      Note Content *
    </label>
    <textarea
      cols="66"
      id="noteContent"
      name="noteContent"
      onChange={[Function]}
      placeholder="Note it up!"
      rows="3"
      type="text"
    />
  </div>
  <div>
    <label
      htmlFor="folderId"
    >
      Folder Association
    </label>
    <select
      id="folderId"
      name="folderId"
      onChange={[Function]}
    />
  </div>
  <div
    className="addNote__button__group"
  >
    <a
      href="/"
      onClick={[Function]}
    >
      <button
        type="reset"
      >
        Cancel
      </button>
    </a>
    <button
      disabled="Note name is required"
      type="submit"
    >
      Save
    </button>
  </div>
</form>
`;
